README.md
# Acoustic Modeler

Allows the user to import sound files and displays information about its waveform, frequencies, and resonance. If you want to observe how different properties of sound interact, this is your module.

---

## Buttons

### Import
Opens a prompt that allows the user to select a file. Currently, only wav and mp3 files are accepted.

### Low/Mid/High
Changes the RT60 graph between low (~200 Hz), mid (~1000 Hz), and high (~5000 Hz) frequencies.

### Split/Merged
Changes whether the RT60 graphs display separately or together.

---

## Graphs

### Waveform
The amplitude of the audio over the file's duration.

### Spectrogram
The intensity of different frequencies in the audio over the file's duration.

### RT60
The intensity of the audio at a certain frequency over the file's duration.

Target frequency can be toggled (see Low/Mid/High), or all three options can be displayed at once (see Split/Merged).

---

## Other Outputs

### Duration
Duration of the audio file in seconds.

### Resonance
Frequency of the highest amplitude in Hertz.

### Difference (currently broken)
Difference between average RT60 and 0.5 seconds.


CHANGELOG.md
# Change Log

## 1.1.0
- Overhauled the file structure and commits
- Changed "Length" to "Duration" in GUI
  - Any references to this sidebar info in README have also been changed to reflect the new name


## 1.0.0
- Added:
  - "Resonance" and "Difference" info in sidebar
  - RT60 graph(s)
  - "Low/Mid/High" and "Split/Merged" buttons for RT60 graph


## 0.2.0
- Added:
  - "Length" info in sidebar
  - Waveform graph
  - Spectrogram graph


## 0.1.0
- Added:
  - "Import" button
  - File cleanup for imports

src/gg_acoustic_modeler/display.py
from tkinter import ttk
from PIL import Image, ImageTk


class Display(ttk.Frame):
    def __init__(self, parent):
        super().__init__(parent)

        # message
        self.message_label = ttk.Label(self, text='', foreground='red')
        self.message_label.grid(row=1, column=0, columnspan=11)

        # save button
        self.import_button = ttk.Button(self, text='Import', command=self.import_clicked)
        self.import_button.grid(row=2, column=0, padx=10)

        # switch low mid high button
        self.switch_button = ttk.Button(self, text='Low', command=self.plot_clicked)
        self.switch_button.grid(row=3, column=0, padx=10)

        # switch merging button
        self.merge_button = ttk.Button(self, text='Split', command=self.merge_clicked)
        self.merge_button.grid(row=4, column=0, padx=10)

        self.length_label = ttk.Label(self, text='')
        self.length_label.grid(row=5, column=0, padx=10)

        self.max_res_label = ttk.Label(self, text='')
        self.max_res_label.grid(row=6, column=0, padx=10)

        self.difference = ttk.Label(self, text='')
        self.difference.grid(row=7, column=0, padx=10)

        self.waveform = ttk.Label(self)
        self.waveform.grid(row=2, column=1, columnspan=30, rowspan=30)

        self.spec = ttk.Label(self)
        self.spec.grid(row=2, column=31, columnspan=30, rowspan=30)

        self.rt60 = ttk.Label(self)
        self.rt60.grid(row=2, column=61, columnspan=30, rowspan=30)

        # set the controller
        self.controller = None

    def set_controller(self, controller):
        """
        Set the controller
        :param controller:
        :return:
        """

        self.controller = controller

    def import_clicked(self):
        """
        Handle import button click event
        :return:
        """

        if self.controller:
            self.controller.import_wav()

    def plot_clicked(self):
        """
        Handle plot button click event
        :return:
        """

        if self.controller:
            self.controller.change_plot_shown()

    def merge_clicked(self):
        """
        Handle merge button click event
        :return:
        """

        if self.controller:
            self.controller.change_plot_combined()

    def show_error(self, message):
        """
        Show an error message
        :param message:
        :return:
        """

        self.message_label['text'] = message
        self.message_label['foreground'] = 'red'
        self.message_label.after(3000, self.hide_message)

    def show_success(self, message):
        """
        Show a success message
        :param message:
        :return:
        """

        self.message_label['text'] = message
        self.message_label['foreground'] = 'green'
        self.message_label.after(3000, self.hide_message)

    def update_length(self, length):
        """
        updates the displayed length
        :param length:
        :return:
        """

        self.length_label['text'] = f"Duration: {length} s"

    def update_max_res(self, max_res):
        """
        updates the displayed max resonance
        :param max_res:
        :return:
        """

        self.max_res_label['text'] = f"Resonance: {max_res} Hz"

    def update_diff(self, diff):
        """
        updates the displayed difference
        :param diff:
        :return:
        """

        self.difference['text'] = f"Difference: {diff} s"

    def update_waveform(self):
        """
        updates the displayed waveform
        :param:
        :return:
        """

        new_image = ImageTk.PhotoImage(Image.open("waveform.png").resize((360, 240)))
        self.waveform.configure(image=new_image)
        self.waveform.image = new_image

    def update_spec(self):
        """
        updates the displayed spectrogram
        :param:
        :return:
        """

        new_image = ImageTk.PhotoImage(Image.open("spec.png").resize((360, 240)))
        self.spec.configure(image=new_image)
        self.spec.image = new_image

    def update_plot(self, plot, combined, name, imported):
        """
        updates the plot and merge buttons, plus the rt60 plot
        :param plot:
        :param combined:
        :param name:
        :param imported:
        :return:
        """

        name_1 = 'N/A'

        match plot:
            case 0:
                name_1 = 'Low'
            case 1:
                name_1 = 'Mid'
            case 2:
                name_1 = 'High'

        self.switch_button['text'] = name_1

        name_2 = 'N/A'

        match combined:
            case False:
                name_2 = 'Split'
            case True:
                name_2 = 'Merged'

        self.merge_button['text'] = name_2

        filename = ''
        if not combined:
            filename = name[plot]
        else:
            filename = 'rt60_combined.png'

        if imported:
            new_image = ImageTk.PhotoImage(Image.open(filename).resize((360, 240)))
            self.rt60.configure(image=new_image)
            self.rt60.image = new_image

    def hide_message(self):
        """
        Hide the message
        :return:
        """

        self.message_label['text'] = ''




src/gg_acoustic_modeler/controller.py
class Controller:
    def __init__(self, model, display):
        self.model = model
        self.display = display

    def import_wav(self):
        """
        Sets the file to import
        :return:
        """

        try:
            # import the model
            self.model.import_wav()
            # show a success message
            self.display.show_success(f'File {self.model.filepath} has been imported.')
            self.model.imported = True
            self.display.update_length(self.model.length)
            self.display.update_max_res(self.model.max_res)
            self.display.update_diff(self.model.avg_diff)
            self.display.update_waveform()
            self.display.update_spec()
            self.display.update_plot(self.model.plot_shown, self.model.plot_combined, self.model.plot_name,
                                     self.model.imported)
        except ValueError as error:
            # show an error message
            self.display.show_error(error)

    def change_plot_shown(self):
        if self.model.plot_shown < 2:
            self.model.plot_shown += 1
        else:
            self.model.plot_shown = 0

        self.display.update_plot(self.model.plot_shown, self.model.plot_combined, self.model.plot_name,
                                     self.model.imported)

    def change_plot_combined(self):
        self.model.plot_combined = not self.model.plot_combined

        self.display.update_plot(self.model.plot_shown, self.model.plot_combined, self.model.plot_name,
                                 self.model.imported)



src/gg_acoustic_modeler/model.py
from tkinter import filedialog as fd
from pydub import AudioSegment
from scipy.io import wavfile
from scipy.signal import welch
import matplotlib.pyplot as plt
import numpy as np


class Model:
    def __init__(self, filepath):
        self.imported = False
        self.filepath = filepath
        self.length = ''
        self.time = None
        self.max_res = None
        self.data_rt60 = None
        self.avg_diff = None
        self.plot_name = ["rt60_low.png", "rt60_mid.png", "rt60_high.png"]
        self.plot_color = ["red", "green", "blue"]
        self.plot_shown = 0
        self.plot_combined = False

    def import_wav(self):
        """
        Import the audio and assign values
        :return:
        """

        filetypes = (
            ('Wave', '*.wav'),
            ('MP3 - requires conversion', '*.mp3'),
            ('All files', '*.*')
        )

        filepath = fd.askopenfilename(
            title='Choose file to import',
            initialdir='/',
            filetypes=filetypes
        )

        wav = AudioSegment.empty()

        if filepath.endswith('.wav'):
            wav = AudioSegment.from_file(filepath, format="wav")
        elif filepath.endswith('.mp3'):
            wav = AudioSegment.from_mp3(filepath)
        else:
            raise ValueError(f'Unsupported file type: {filepath}')

        self.filepath = filepath

        wav.tags = None
        mono_wav = wav.set_channels(channels=1)
        mono_wav.export("acoustics.wav", format="wav")

        sample_rate, data = wavfile.read("acoustics.wav")

        self.length = round(data.shape[0] / sample_rate, 2)

        frequencies, power = welch(data, sample_rate, nperseg=4096)
        self.max_res = round(frequencies[np.argmax(power)])

        plt.plot(np.linspace(0., self.length, data.shape[0]), data)
        plt.title("Waveform")
        plt.xlabel("Time (s)")
        plt.ylabel("Amplitude")
        plt.grid()
        plt.savefig("waveform.png")
        plt.clf()

        spectrum, freqs, t, im = plt.specgram(data, Fs=sample_rate, NFFT=1024,
                                              cmap=plt.get_cmap('autumn_r'))
        cbar = plt.colorbar(im)
        plt.title("Spectrogram")
        plt.xlabel("Time (s)")
        plt.ylabel("Frequency (Hz)")
        cbar.set_label("Intensity (dB)")
        plt.grid(color="red")
        plt.savefig("spec.png")
        plt.clf()

        self.data_rt60 = 0.0

        for x in range(0, 2):
            for y in range(0, 3):
                self.data_rt60 += self.plot_rt60(x, y, spectrum, freqs, t, self.plot_color[y])

                if x == 0:
                    plt.savefig(self.plot_name[y])
                    plt.clf()

            if x == 1:
                plt.savefig("rt60_combined.png")
                plt.clf()

        self.avg_diff = (self.data_rt60 / 6) - 0.5

    @staticmethod
    def plot_rt60(x, y, spectrum, freqs, t, color):
        def find_nearest_value(array, value):
            array = np.asarray(array)
            idx = (np.abs(array - value)).argmin()
            return array[idx]

        def_freq = (1000 * pow(5, y - 1))
        target_freq = def_freq
        for target_freq in freqs:
            if target_freq > def_freq:
                break

        data_db = (10 * np.log10(spectrum[np.where(freqs == target_freq)]))[0]

        if x == 0 or y == 0:
            plt.figure()
            plt.grid()

        plt.plot(t, data_db, linewidth=1, alpha=0.7, color=color)
        plt.title("RT60")
        plt.xlabel("Time (s)")
        plt.ylabel("Power (dB)")

        max_index = np.argmax(data_db)
        max_value = data_db[max_index]
        plt.plot(t[max_index], data_db[max_index], 'k^')

        data_slice = data_db[max_index:]

        max_less_5_value = find_nearest_value(data_slice, max_value - 5)
        max_less_5_index = np.where(data_db == max_less_5_value)
        plt.plot(t[max_less_5_index], data_db[max_less_5_index], 'ks')

        max_less_25_value = find_nearest_value(data_slice, max_value - 25)
        max_less_25_index = np.where(data_db == max_less_5_value)
        plt.plot(t[max_less_5_index], data_db[max_less_5_index], 'kp')

        rt20 = t[max_less_5_index] - t[max_less_25_index]
        rt60 = 3 * rt20
        return float(rt60)

